### ISSUE TITLE
NC upgrade erased entire data folder content

### ISSUE THREAD

**Issue by MarcS1975** [nextcloud/docker#2132]
I have recently tried to upgrade from 27.0.1.2 to 28.0.1.1 and this action erased all content of my data folder (which was mapped as a docker volume). Luckily I had a backup but I am curious why the NC upgrade procedure did not retain my old data. 
You can see from my compose file below, that I have mapped an existing folder 'MEDIA' into the NC container. This MEDIA folder held 1TB of data has been filled prior to installing NC.

`- /srv/mypath/mydata/MEDIA:/var/www/html/media`

I add MEDIA as External Folder in Nextcloud with file type "local" and access to all data worked fine. 
When I decided to upgrade NC, I added **:latest** behind the NC container name and re-deployed the stack (in Portainer) with the option "pull containers". 

Then the NC log files started filing up with the following recurring errors:

```
28/12/2023
02:06:09
Initializing nextcloud 28.0.1.1 ...
28/12/2023
02:06:09
Upgrading nextcloud from 27.0.1.2 ...
28/12/2023
02:06:10
Nextcloud or one of the apps require upgrade - only a limited number of commands are available
28/12/2023
02:06:10
You may use your browser or the occ upgrade command to do the upgrade
28/12/2023
02:06:10
rsync: [generator] delete_file: rmdir(media) failed: Device or resource busy (16)
28/12/2023
02:06:32
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1338) [sender=3.2.7]
```

Any ideas why the upgrade resulted in all my data erased (but not the folder itself). Is NC not supposed to keep all persistent data untouched during an upgrade?

My docker compose file is as follows
```
version: '2'

volumes:
  nextcloud:
  db:

services:
  db:
    image: mariadb:10.6
    restart: always
    command: --transaction-isolation=READ-COMMITTED --log-bin=binlog --binlog-format=ROW
    volumes:
      - /srv/mypath/NC_db:/var/lib/mysql
    environment:
      - MYSQL_ROOT_PASSWORD=xxx
      - MYSQL_PASSWORD=yyy
      - MYSQL_DATABASE=xxx
      - MYSQL_USER=yyy
      - REDIS_HOST=redis
      - REDIS_HOST_PASSWORD=xxx

  app:
    image: nextcloud**:latest**
    restart: always
    ports:
      - 8585:80

    links:
      - db
    volumes:
      - /srv/mypath/NC_app:/var/www/html
      **- /srv/mypath/mydata/MEDIA:/var/www/html/media**
      
    environment:
      - MYSQL_PASSWORD=xxx
      - MYSQL_DATABASE=yyy
      - MYSQL_USER=xxx
      - MYSQL_HOST=db
      - PHP_MEMORY_LIMIT=1024M
      - PHP_UPLOAD_LIMIT=7G

  cron:
    image: rcdailey/nextcloud-cronjob
    restart: always
    network_mode: none
    depends_on:
    - app
    volumes:
    - /var/run/docker.sock:/var/run/docker.sock:ro
    - /etc/localtime:/etc/localtime:ro
    environment:
    - NEXTCLOUD_CONTAINER_NAME=nc40-app-1
    - NEXTCLOUD_PROJECT_NAME=
    - NEXTCLOUD_CRON_MINUTE_INTERVAL=5
```

**Comment by martadinata666** [comment]
The path you mount is not in upgrade.exclude. https://github.com/nextcloud/docker/blob/master/upgrade.exclude

**Comment by MarcS1975** [comment]
thanks Dedy. Does that mean any folder not in upgrade.exclude gets erased
during an upgrade?

**Comment by martadinata666** [comment]
In simple terms yes. `/var/www/html/data/media` is your best bet or use nextcloud external storage plugin

**Comment by MarcS1975** [comment]
Have you used External?.
I tried the External storage plug in smbclient but it does not work.
Constantly throws errors in the NC log file.

**Comment by martadinata666** [comment]
tldr; yes, it works.

This docker image is not `complete` packages that needed to run out of the box missing, I dunno why it doesn't want to add samba support and expected users to build one ourself.

https://github.com/nextcloud/docker/tree/master/.examples/dockerfiles/smb

**Comment by joshtrichards** [comment]
See https://github.com/nextcloud/docker#persistent-data - last couple paragraphs explains the behavior you encountered.

**Comment by joshtrichards** [comment]
Check #2133 for a proposed adjustment to the README to make this more prominent and to clarify the existing language. Feedback welcome.

**Comment by MarcS1975** [comment]
ok thanks. and how can I amend the file update.exclude?

**Comment by MarcS1975** [comment]
thanks - I have done the smbclient addition but I still constantly get
errors in the log files:

[PHP] Error: Undefined array key "attributes" at
/var/www/html/apps/files_external/3rdparty/icewind/smb/src/Wrapped/Parser.php#155
from ? by -- at 28 Dec 2023, 14:00:37
[files] Error: Malformed state response from server
from ? by -- at 28 Dec 2023, 13:54:22

**Comment by joshtrichards** [comment]
> thanks - I have done the smbclient addition but I still constantly get
errors in the log files:

> [PHP] Error: Undefined array key "attributes" at
/var/www/html/apps/files_external/3rdparty/icewind/smb/src/Wrapped/Parser.php#155
from ? by -- at 28 Dec 2023, 14:00:37
[files] Error: Malformed state response from server
from ? by -- at 28 Dec 2023, 13:54:22

That is either an upstream bug (which I'm fairly certain there is an existing issue open in the `server` repo for that) *or* a compatibly issue with some combination of the server-side/OS SMB related components and your SMB server (could even be both). It may or may not be a bug in Nextcloud - there are many SMB protocol/library version combinations.

**Comment by MarcS1975** [comment]
It may well be an upstream bug but the impact on Nextcloud inside Docker is
more significant because NC Docker users are currently left with only 1
alternative to utilise existing external data, via mounting into the
countainer. And that creates its own set of problems related to permissions.
I wish that SMB integration into NC could be fixed. It has been going for
at least 2 years (maybe longer).
